smack-core/src/main/java/org/jivesoftware/smack/altConnections/HttpLookupMethod.java

```diff
@@ -0,0 +1,218 @@
+/**
+ *
+ * Copyright 2020 Aditya Borikar.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.jivesoftware.smack.altConnections;
+
+import java.io.ByteArrayOutputStream;
+import java.io.IOException;
+import java.io.InputStream;
+import java.net.HttpURLConnection;
+import java.net.URI;
+import java.net.URISyntaxException;
+import java.net.URL;
+import java.util.ArrayList;
+import java.util.Iterator;
+import java.util.List;
+
+import org.jivesoftware.smack.util.MultiMap;
+import org.jivesoftware.smack.util.PacketParserUtils;
+import org.jivesoftware.smack.xml.XmlPullParser;
+import org.jivesoftware.smack.xml.XmlPullParserException;
+
+import org.jxmpp.jid.DomainBareJid;
+
+/**
+ * <p><i>The HTTP lookup method uses web host metadata to list the URIs of alternative connection methods.</i></p>
+ *
+ * <p>In order to obtain host-meta XRD element from the host in the form of an <i>InputStream</i>,
+ * use {@link #getXrdStream(DomainBareJid)} method. To obtain endpoints for Bosh or Websocket
+ * connection endpoints from host, use {@link LinkRelation#BOSH} and {@link LinkRelation#WEBSOCKET}
+ * respectively with the {@link #lookup(DomainBareJid, LinkRelation)} method. In case one is looking
+ * for endpoints described by other than BOSH or Websocket LinkRelation, use the more flexible
+ * {@link #lookup(DomainBareJid, String)} method.</p>
+ * Example:<br>
+ * <pre>
+ * {@code
+ * List<URI> endpoints = HttpLookupMethod.lookup(DomainBareJid, LinkRelation.WEBSOCKET);
+ * }
+ * </pre>
+ * @see <a href="https://xmpp.org/extensions/xep-0156.html#http">
+ *     HTTP Lookup Method from XEP-0156.
+ *     </a>
+ */
+public final class HttpLookupMethod {
+
+    private static final int BUFFER_SIZE = 256;
+    private static final String XRD_NAMESPACE = "http://docs.oasis-open.org/ns/xri/xrd-1.0";
+
+    private static int contentLength;
+    private static HttpURLConnection httpURLConnection;
+    private static MultiMap<DomainBareJid, MultiMap<String, URI>> CACHE = new MultiMap<>();
```

Comment:
Caches are never simple maps. As this would allow them to grow indefinitely. I am still not sure if we really need a cache, as I'd epxect that the HTTP library will do caching for you. I'd say we remove the cache for now and re-introduce it if we see a need for it.